### ISSUE TITLE
Support S3 additional checksums in high-level S3 commands

### ISSUE THREAD

**Issue by m-radzikowski** [aws/aws-cli#6750]
**Is your feature request related to a problem? Please describe.**

Newly released [additional S3 checksums](https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/) feature enhances the SDKs operations by calculating selected checksum value on file upload. This also includes multipart upload. However, this new feature is not present in the high-level S3 commands.

**Describe the solution you'd like**

`--checksum-algorithm` parameter in the `aws s3` commands, especially in the `aws s3 cp`.

**Describe alternatives you've considered**

Using low-level commands.

**Comment by tim-finnigan** [comment]
Hi @m-radzikowski thanks for the feature request. I’m going to bring this up for discussion with the team and will post an update here.

**Comment by rajivnarayan** [comment]
This would be a useful addition to the high-level commands. For reference here is a solution using s3api:
Would have been nice if MD5 digests were included as an option.

```
# aws-cli version 2.7.16
# https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/

# compute and save checksum on upload / copy
# algorithms supported: crc32 crc32c sha1 sha256
aws s3api put-object --body <file_name> --checksum-algorithm crc32 --bucket <bucket_name> --key <key_name>

# retrieve the checksum```
# ChecksumCRC32 ChecksumCRC32C ChecksumSHA1 ChecksumSHA256
aws s3api head-object --bucket <bucket_name> --key <key_name> --checksum-mode Enabled --Query ChecksumCRC32 --output text
```

**Comment by jonathansampson** [comment]
+1 to support for checksums when syncing.

**Comment by saksham** [comment]
+1 on this feature.

**Comment by genvidkyle** [comment]
+1

**Comment by ashepherd** [comment]
+1

**Comment by MaksymSimchuk-prxt** [comment]
+1

**Comment by khilnani** [comment]
+1

**Comment by sarthakjain271095** [comment]
@rajivnarayan i stumbled upon [this](https://github.com/aws/aws-cli/issues/6750#issuecomment-1195959947). The sha256 checksum value being returned from aws doesn't seem to be right. Additionally, as per the [cli doc](https://docs.aws.amazon.com/cli/latest/reference/s3api/put-object.html), `--checksum-algorithm` param is only supported when using SDK. Have you faced issues about the sha256 value not being rightly calculated?

**Comment by rajivnarayan** [comment]
Works fine for me with the latest aws cli (2.12.16). Note that the checksum
in base64 encoded as detailed here:
https://aws.amazon.com/getting-started/hands-on/amazon-s3-with-additional-checksums/?ref=docs_gateway/amazons3/checking-object-integrity.html

```
BUCKET=my-test-bucket
KEY=hello_checksum.txt
echo "Hello world!" > hello.txt

# Compute base64 encoded sha256
shasum -a 256 hello.txt|cut -f1 -d\ |xxd -r -p|base64
# C6kE6uh3O3DHUzPbTeLzrEWorU3bobJC8LPPwZk5Hdg=

# compute and save checksum on upload / copy
# algorithms supported: crc32 crc32c sha1 sha256
aws s3api put-object --body hello.txt --checksum-algorithm sha256 --bucket
${BUCKET} --key ${KEY}

# retrieve the checksum
aws s3api head-object --bucket ${BUCKET} --key ${KEY} --checksum-mode
Enabled --query ChecksumSHA256 --output text
# C6kE6uh3O3DHUzPbTeLzrEWorU3bobJC8LPPwZk5Hdg=
```

**Comment by Park-minkyu** [comment]
I believe that most of use-cases are probably using high-level s3 command s3 cp or sync.
can we have more information to think through the implementation?

**Comment by YoongLoong** [comment]
Hi, i have a situation in my usage on aws s3 sync:
File A: abcdefg.txt (1234 Bytes) (md5sum: a954ea6b12db7d788350b4c008ee26fb)
File B: abcdefg.txt (1234 Bytes) (md5sum: bd84bb07f07a58e832bb3c72e3db4663)

After I aws s3 sync the File A to the s3 location and try to sync the File B to the same s3 location it will skipped. Sync is only taking care of **"filesize change"** only instead of **md5sum change** which defeat the purpose of "SYNC" and it will give the wrong impression that there is no change between File A and File B. Would this be fixed or taken out as it is confusing to the user that they have the identical files being "sync"?

**Comment by YoongLoong** [comment]
I do support the changes in high level implementation aws s3 sync command however this feature should be disabled temporary when it is not being fixed at the moment. We have no idea when will this "new" feature exist (the thread had been 1 year plus) but the "sync" command is misleading the user that they have "sychronized" the files while it is not always the case. It may caused the financial lost to the company if the "wrong" object had been synchronized. I am forced to do the workaround to fix this aws s3 sync issue to ensure the "different md5sum with same file size" file being uploaded (skipped using aws s3 sync at the moment).

**Comment by YoongLoong** [comment]
May I have the update on the aws sync bug issue? This is causing a lot of inconvenience to sync the file(s) from AWS S3 now.

**Comment by aemous** [comment]
This feature has been released into version 2.18.0. Closing issue.
